Automated readability index (#181)
exercism/awk: exercises/practice/automated-readability-index/.docs/instructions.md

# Instructions

As we grow up, we can understand more and more complex texts. 
But how is the level of complexity of a given text assessed, 

> and how can we make programs which do that? In this exercise, 

Reviewer: [Exercism's style guide for markdown](https://exercism.org/docs/building/markdown/markdown) uses one sentence per line. I resisted at first but have come to see the wisdom behind this style.

Author: fixed

you will write a program which will determine how difficult a 
text is to read and what age it is best suited for.

We will use the [Automated readability index (ARI)][wiki] formula 

> for calculating the text complexity. The ARI was developed in 1967 

Reviewer 2: ```suggestion
for calculating text complexity. The ARI was developed in 1967 
```

> for calculating text complexity. The ARI was developed in 1967 
> and assesses the U.S. grade level required to read a piece of text. 

Reviewer: ```suggestion
for calculating text complexity. The ARI was developed in 1967 
to assess the U.S. grade level required to read a piece of text. 
```

It does this by measuring the complexity of words and sentences. 
Complexity is measured as characters per word and words per sentence. 
A higher value is more complex.

## Text Preview

> Your program should display the first three sentences of the text, 
> but no more than 50 characters. If the first three sentences exceed 
> 50 characters, truncate them and add an ellipsis at the end. Otherwise, 
> if the text is more than three sentences, put an ellipsis on the 
> fourth line.

Reviewer: ```suggestion
Your program should display the first three sentences of the text, but no more than 50 characters for each sentence.
If any sentence exceeds 50 characters, print only the first 50 characters followed by an ellipsis (`...`).
If the text contains more than three sentences, output only an ellipsis as the fourth line.
```

> Suppose there is a file named "ari.txt" with the following contents:

Reviewer: ```suggestion
Suppose there is a file named "ari.txt" with the following contents (this text will be used for all examples below):
```

> ARI was designed for military use in 1967. It was for real-time 
> monitoring of readability on typewriters. It offered an automated 
> counting method. It has similar results to other frameworks.
> This validates its accuracy.

Then the program should print:

```text
The text is:
ARI was designed for military use in 1967.
It was for real-time monitoring of readability on ...
It offered an automated counting method.
...
```

## Text Statistics

The program should calculate and print the number of characters, words
and sentences in the text. Sentences can end with a full stop, 

> an exclamation mark, or a question mark. The last sentence may end with 
> or without a punctuation mark. Whitespaces separate words in a sentence. 

Reviewer: ```suggestion
an exclamation mark, or a question mark. 
Whitespaces separate words in a sentence. 
```
No tests are missing the trailing puncutation. No need to add the extra complexity to user's solutions.

When counting the number of characters per word, count all non-blank 
characters, including punctuation marks. Print the statistics after 
the text preview, separated by an empty line.

> Here is an example output:
> 
> ```text
> The text is:
> ARI was designed for military use in 1967.
> It was for real-time monitoring of readability on ...
> It offered an automated counting method.
> ...
> 
> Words: 34
> Sentences: 5
> Characters: 186
> ```

Reviewer: ````suggestion
Here is an example output for just the text statistics part:

```text
Words: 34
Sentences: 5
Characters: 186
```
````

## Automated readability index

Given these details about the text, we can calculate 
the level of its complexity using the following formula:

Score = 4.71 * chars / words + 0.5 * words / sentences - 21.43

The program should calculate and print the score and the approximate age needed to comprehend the text. 

| Score | Age	  | Grade Level     |
|-------|-------|-----------------|
| 1     | 5-6   | Kindergarten    |
| 2     | 6-7   | First Grade     |
| 3     | 7-8   | Second Grade    |
| 4     | 8-9   | Third Grade     |
| 5     | 9-10  | Fourth Grade    |
| 6     | 10-11 | Fifth Grade     |
| 7     | 11-12 | Sixth Grade     |
| 8     | 12-13 | Seventh Grade   |
| 9     | 13-14 | Eighth Grade    |
| 10    | 14-15 | Ninth Grade     |
| 11    | 15-16 | Tenth Grade     |
| 12    | 16-17 | Eleventh Grade  |
| 13    | 17-18 | Twelfth Grade   |
| 14    | 18-22 | College student |

The score should be printed with two significant digits after 

> the decimal point. When used used to find the age or grade level, the score should be rounded

Reviewer 2: ```suggestion
the decimal point. When used to find the age or grade level, the score should be rounded
```

The line wrapping can use an update here.

# Instructions

> As we grow up, we can understand more and more complex texts. But how is the level of complexity of a given text assessed, and how can we make programs which do that? 

Reviewer 2: Please wrap sentences consistently :)

If you're wrapping at column offsets, you may need to reformat paragraphs every time they change.
If you wrap by sentences, it tends to be less work, but less uniform width.

In this exercise, you will write a program which will determine how difficult a text is to read
and what age it is best suited for.

We will use the
[Automated readability index (ARI)][wiki] formula for calculating the text complexity.
The ARI was developed in 1967 and assesses the U.S. grade level required to read 

> a piece of text. In some ways, it is similar to other formulas. Its main difference 

Reviewer 2: > In some ways, it is similar to other formulas.

Does this sentence add anything? Can it be dropped?

Author: Some exercises have two files: `introduction.md` and `instruction.md`. I think this text fragment may be excluded from instruction, and if we would like to create an introduction, it can be added here. It is a quote from an article.

Reviewer 2: > In some ways, it is similar to other formulas. Its main difference is rather than counting syllables, it counts characters.

can simply be:

> Its main difference from other formulas is rather than counting syllables, it counts characters.

Author: fixed

is rather than counting syllables, it counts characters. The more characters per word, 

> the harder the word. It also counts the number of words per sentences. This sets it apart from some 

Reviewer 2: Presumable, "the more words per sentence, the harder the sentence"?
Maybe combine both sentences? Something along the lines of, "It counts the number of characters per word and words per sentence. The higher the count, the more complex the text is considered."

Author: I copied this piece of text from one article. Because English is not my native language, I am not very good at writing it. If it is possible to rewrite this fragment with other words, then I think this should be done.

Reviewer 2: If I were entirely rewriting it, I might suggest something like:

> We will use the [Automated readability index (ARI)][wiki] formula for calculating the text complexity.
> The ARI was developed in 1967 and assesses the U.S. grade level required to read 
a piece of text. It does this by measuring the complexity of words and sentences. Complexity is measured as characters per word and words per sentence. A higher value is more complex.

How does that sound?

Author: It sounds good to me. 
Do we need `the` before `text` in the first sentence?

Author: fixed

other formulas.

## Text Preview

Your program should display the first three sentences of the text, but no more than 50 characters. If the first three sentences exceed 50 characters, truncate them and add an ellipsis at the end. Otherwise, if the text is more than three sentences, put an ellipsis on the fourth line.

Suppose the text starts with the following lines:

> ARI was designed for military use in 1967. It was for real-time 
> monitoring of readability on typewriters. It offered an automated 
> counting method. It has similar results to other frameworks.
> This validates its accuracy.

Then the program should print:

```text
The text is:
ARI was designed for military use in 1967.
It was for real-time monitoring of readability on ...
It offered an automated counting method.
...
```

## Text Statistics

The program should calculate and print the number of sentences, words 

> and characters in the text. The sentences can end with a full stop, 

Reviewer 2: Does it make sense to go from small items to large? Characters, words, sentences.

Author: The google request did not give a clear result for "count characters, words, sentences."

> Counting words, letters, sentences, characters
https://kalkulatori.lv/en/word-count

The listing order is not important to me.

Reviewer 2: I would suggest going from small to large here.

Author: fixed

> The program should calculate and print the number of sentences, words 
> and characters in the text. The sentences can end with a full stop, 

Reviewer 2: ```suggestion
The program should calculate and print the number of sentences, words 
and characters in the text. Sentences can end with a full stop, 
```

an exclamation mark, or a question mark. The last sentence may end with or 

> without a punctuation mark. Spaces separate words in a sentence. 

Reviewer 2: Does "Spaces" exclude other whitespace like tabs? Those are non-blank so would they be included in the number of chars per word?

Author: no, it means any `[:space:]` chars.

Reviewer 2: Then "whitespace" should be used.

Author: fixed

When counting the number of characters per word, count all non-blank characters, including punctuation marks. Print 

> the statistics after the text preview, separating them with an empty line.

Reviewer 2: ```suggestion
the statistics after the text preview, separated by an empty line.
```

Here is an example output:

```text
The text is:
ARI was designed for military use in 1967.
It was for real-time monitoring of readability on ...
It offered an automated counting method.
...

Words: 34
Sentences: 5
Characters: 186
```

## Automated readability index

Given these details about the text, we can calculate 
the level of its complexity using the following formula:

```math

> Score = 4.71 ( { characters \over words }) + 0.5 ({words \over sentences}) - 21.43

Reviewer 2: Does this render?

Author: No idea. It is rendered in GitHub markdown, but it doesn't render in the Exercism forum. So I'm afraid it may not be rendered.

Reviewer 2: Sasha responded on the forum. There's a request for this, but it is not yet supported.

As such, we should drop the math markdown and stick with something more programming-looking, e.g. `score = 4.71 * chars / words + 0.5 * words / sentences - 21.43`

Author: Fixed

```

The program should calculate and print the score and the approximate age needed to comprehend the text.

| Score | Age	  | Grade Level     |
|-------|-------|-----------------|
| 1     | 5-6   | Kindergarten    |
| 2     | 6-7   | First Grade     |
| 3     | 7-8   | Second Grade    |
| 4     | 8-9   | Third Grade     |
| 5     | 9-10  | Fourth Grade    |
| 6     | 10-11 | Fifth Grade     |
| 7     | 11-12 | Sixth Grade     |
| 8     | 12-13 | Seventh Grade   |
| 9     | 13-14 | Eighth Grade    |
| 10    | 14-15 | Ninth Grade     |
| 11    | 15-16 | Tenth Grade     |
| 12    | 16-17 | Eleventh Grade  |
| 13    | 17-18 | Twelfth Grade   |
| 14    | 18-22 | College student |

> As an example, suppose there is a file named "ari.txt" with the following contents:

Reviewer 2: Is this a duplicate of the earlier block?

Author: fixed

Here is an example output:

```text
The text is:
ARI was designed for military use in 1967.
It was for real-time monitoring of readability on ...
It offered an automated counting method.
...

Words: 34
Sentences: 5
Characters: 186
```

## Automated readability index

Given these details about the text, we can calculate 
the level of its complexity using the following formula:

```math
Score = 4.71 ( { characters \over words }) + 0.5 ({words \over sentences}) - 21.43
```

The program should calculate and print the score and the approximate age needed to comprehend the text. 

| Score | Age	  | Grade Level     |
|-------|-------|-----------------|
| 1     | 5-6   | Kindergarten    |
| 2     | 6-7   | First Grade     |
| 3     | 7-8   | Second Grade    |
| 4     | 8-9   | Third Grade     |
| 5     | 9-10  | Fourth Grade    |
| 6     | 10-11 | Fifth Grade     |
| 7     | 11-12 | Sixth Grade     |
| 8     | 12-13 | Seventh Grade   |
| 9     | 13-14 | Eighth Grade    |
| 10    | 14-15 | Ninth Grade     |
| 11    | 15-16 | Tenth Grade     |
| 12    | 16-17 | Eleventh Grade  |
| 13    | 17-18 | Twelfth Grade   |
| 14    | 18-22 | College student |

> Please note that the Score must be printed with two significant digits after 
> the decimal point, but when used as a table index, you must round this value 
> to the nearest whole number but not more than 14. Below is a sample output of 
> the program:

Reviewer 2: ```suggestion
The score should be printed with two significant digits after 
the decimal point. When used used to find the age or grade level, the score should be rounded
to the nearest whole number and capped at 14. Below is a sample output of 
the program:
```

```text
The text is:
ARI was designed for military use in 1967.
It was for real-time monitoring of readability on ...
It offered an automated counting method.
...

Words: 34
Sentences: 5
Characters: 186

> The score is: 7.74

Reviewer 2: This line looks like a sentence. A `:` is not needed in a sentence. This line is also inconsistent with the prior three lines.

Consider making the format uniform across all the metrics.

Author: fixed

# Instructions

> As we grow up, we can understand more and more complex texts. But how to assess 
> the level of complexity of a given text, and how to teach the computer to do so? 
> In this exercise, you will write a program determining how difficult a text is 
> and for what age it is best suited.

Reviewer 2: ```suggestion
As we grow up, we can understand more and more complex texts. But how is the level of complexity of a given text assessed, and how can we make programs which do that? 
In this exercise, you will write a program which will determine how difficult a text is to read
and what age it is best suited for.
```

> For a formula for calculating the complexity, we will use 
> [Automated readability index (ARI)][wiki] 
> developed in 1967 year. The ARI assesses the U.S. grade level required to read 
> a piece of text. In some ways, it is similar to other formulas. Its difference 
> is rather than counting syllables, it counts characters. The more characters, 
> the harder the word. It also counts sentences. This sets it apart from some 
> other formulas.

Reviewer 2: ```suggestion
We will use the
[Automated readability index (ARI)][wiki] formula for calculating the text complexity.
The ARI was developed in 1967 and assesses the U.S. grade level required to read 
a piece of text. In some ways, it is similar to other formulas. Its main difference 
is rather than counting syllables, it counts characters. The more characters per word, 
the harder the word. It also counts the number of words per sentences. This sets it apart from some 
other formulas.
```

Should this link to other assessments?

## Text Preview

> Your program should display the first three sentences of the text. If the sentence 
> is longer than 50 characters, truncate the sentence and add an ellipsis at the end. 
> Put an ellipsis on the fourth line if the text is more than three sentences.

Reviewer 2: ```suggestion
Your program should display the first three sentences of the text, but no more than 50 characters. If the first three sentences exceed 50 characters, truncate them and add an ellipsis at the end. Otherwise, if the text is more than three sentences, put an ellipsis on the fourth line.
```

Suppose the text starts with the following lines:

> Billy always listens to his mother. He always does what
> she says. If his mother says, "Brush your teeth," Billy
> brushes his teeth. If his mother says, "Go to bed," Billy
> goes to bed...

Then the program should print:

```text
The text is:
Billy always listens to his mother.
He always does what she says.
If his mother says, "Brush your teeth," Billy brus...
...
```

## Text Statistics

> The program should calculate and print the exact number of sentences, words 
> and characters in the text. The sentences can end with a full stop, 
> an exclamation mark, and a question mark. The last sentence can be with or 
> without a punctuation mark at the end. Spaces separate words in a sentence; 
> we count all non-blank characters, including punctuation marks. Print 
> the statistics after the text preview, separating them with an empty line, 
> as in the example:

Reviewer 2: ```suggestion
The program should calculate and print the number of sentences, words 
and characters in the text. The sentences can end with a full stop, 
an exclamation mark, or a question mark. The last sentence may end with or 
without a punctuation mark. Spaces separate words in a sentence. 
When counting the number of characters per word, count all non-blank characters, including punctuation marks. Print 
the statistics after the text preview, separating them with an empty line.
Here is an example output:
```

```text
The text is:
ARI was designed for military use in 1967.
It was for real-time monitoring of readability on ...
It offered an automated counting method.
...

Words: 34
Sentences: 5
Characters: 186
```

## Automated readability index

Knowing the main characteristics of the text, we can calculate 
the level of its complexity using the following formula:

```math
Score = 4.71 ( { characters \over words }) + 0.5 ({words \over sentences}) - 21.43
```

The program should calculate and print score as well as approximate age needed to comprehend the text.

| Score | Age	  | Grade Level     |
|-------|-------|-----------------|
| 1     | 5-6   | Kindergarten    |
| 2     | 6-7   | First Grade     |
| 3     | 7-8   | Second Grade    |
| 4     | 8-9   | Third Grade     |
| 5     | 9-10  | Fourth Grade    |
| 6     | 10-11 | Fifth Grade     |
| 7     | 11-12 | Sixth Grade     |
| 8     | 12-13 | Seventh Grade   |
| 9     | 13-14 | Eighth Grade    |
| 10    | 14-15 | Ninth Grade     |
| 11    | 15-16 | Tenth Grade     |
| 12    | 16-17 | Eleventh Grade  |
| 13    | 17-18 | Twelfth Grade   |
| 14    | 18-22 | College student |

> As an example, suppose there is a file named "ari.txt" with the following contents:

Author: I think yes. Do we keep ARI.text?
